core/node/node_storage_init/src/external_node.rs

```diff
@@ -0,0 +1,141 @@
+use std::{sync::Arc, time::Instant};
+
+use anyhow::Context as _;
+use tokio::sync::watch;
+use zksync_block_reverter::BlockReverter;
+use zksync_dal::{ConnectionPool, Core};
+use zksync_health_check::AppHealthCheck;
+use zksync_object_store::ObjectStoreFactory;
+use zksync_reorg_detector::ReorgDetector;
+use zksync_shared_metrics::{SnapshotRecoveryStage, APP_METRICS};
+use zksync_snapshots_applier::{
+    RecoveryCompletionStatus, SnapshotsApplierConfig, SnapshotsApplierTask,
+};
+use zksync_types::{L1BatchNumber, L2ChainId};
+use zksync_web3_decl::client::{DynClient, L2};
+
+use crate::{node_role::NodeRole, SnapshotRecoveryConfig};
+
+#[derive(Debug)]
+pub struct ExternalNodeRole {
+    pub l2_chain_id: L2ChainId,
+    pub client: Box<DynClient<L2>>,
+}
+
+#[async_trait::async_trait]
+impl NodeRole for ExternalNodeRole {
+    /// Will perform genesis initialization if it's required.
+    /// If genesis is already performed, this method will do nothing.
+    async fn genesis(&self, pool: &ConnectionPool<Core>) -> anyhow::Result<()> {
+        let mut storage = pool.connection_tagged("en").await?;
+        zksync_node_sync::genesis::perform_genesis_if_needed(
+            &mut storage,
+            self.l2_chain_id,
+            &self.client.clone().for_component("genesis"),
+        )
+        .await
+        .context("performing genesis failed")
+    }
+
+    async fn is_genesis_performed(&self, pool: &ConnectionPool<Core>) -> anyhow::Result<bool> {
+        let mut storage = pool.connection_tagged("en").await?;
+        let needed = zksync_node_sync::genesis::is_genesis_needed(&mut storage).await?;
+        Ok(!needed)
+    }
+
+    async fn snapshot_recovery(
+        &self,
+        stop_receiver: watch::Receiver<bool>,
+        pool: &ConnectionPool<Core>,
+        recovery_config: SnapshotRecoveryConfig,
+        app_health: Arc<AppHealthCheck>,
```

Comment:
I agree generally, but (same with [this concept](https://github.com/matter-labs/zksync-era/pull/2363#discussion_r1662108602)) do we need this kind of generalization right now? 

If we have it as a single trait, we will need a structure per initialization step, and given that the main node/EN is somewhat different, it will _likely_ result in two sets of structures, which will have significant cognitive complexity. 

Wouldn't it be better to wait until we have two more or less aligned implementations, and then rewrite these interfaces accordingly?

Reply:
Though I agree about the arguments -- it certainly can be made a part of the role implementation.

Reply:
Actually, I thought about the arguments for a while -- I think that passing them explicitly to the abstract interface creates artificial constraints for ourselves, which may be beneficial for the purpose of further unification of EN/main node initialization approaches. IMHO, short-term we can treat it as a feature, not as a bug, and rework the interfaces once we have a common set of components for both EN and main node.